core/classifiers/classifier_utils.py

```diff
@@ -60,15 +61,152 @@ def extract_svm_parameters(clf):
     }
 
 
-def unicode_validator_for_classifier_data(var):
-    """Validates that incoming object contains unicode literal strings."""
-    if isinstance(var, dict):
-        for k in var.keys():
+def unicode_validator_for_classifier_data(classifier_data):
+    """Validates that incoming object contains unicode literal strings.
+
+    Args:
+        classifier_data: *. The trained classifier model data.
+
+    Raises:
+        Exception. If any of the strings in classifier data is not a unicode
+            string.
+    """
+    if isinstance(classifier_data, dict):
+        for k in classifier_data.keys():
             if isinstance(k, str):
                 raise Exception('Expected %s to be unicode but found str.' % k)
-            unicode_validator_for_classifier_data(var[k])
-    if isinstance(var, (list, set, tuple)):
-        for item in var:
+            unicode_validator_for_classifier_data(classifier_data[k])
+    if isinstance(classifier_data, list):
+        for item in classifier_data:
             unicode_validator_for_classifier_data(item)
-    if isinstance(var, str):
-        raise Exception('Expected \'%s\' to be unicode but found str.' % var)
+    if isinstance(classifier_data, str):
+        raise Exception(
+            'Expected \'%s\' to be unicode but found str.' % classifier_data)
+
+
+def make_sure_that_list_has_uniform_structure(data):
+    """Makes sure that list and its sublist/dicts/items have same uniform
```

Comment:
Can you give a more precise definition of "uniform structure"?

For example is `[['a', 'b', 'c'], 'd']` uniform? Based on your algorithm I think it is. But that's not intuitive, especially since `[[['a', 'b', 'c']], 'd']` would not, I think, be considered uniform by the algorithm (because list != str).

Reply:
Uniform as in, the structure is same at every level of depth. So, no, `[['a', 'b', 'c'], 'd']` is not uniform because at next level one item is list and other is string. However, it is possible to modify algorithm to not check uniformity at every level and only check that all leaf nodes have same type, regardless of their individual levels in the tree.

Reply:
OK, I'm happy either way as long as we have a consistent definition that satisfies all three criteria.

Reply:
Yep, all of them are satisfied. Having it this way just adds more restriction to what lists can contain. It says that lists, in addition to having the same type of leaf value, should also have consistent structure at all levels of depth.

Reply:
OK. Yep, I'm fine with that! I don't think your algorithm as it currently stands does that, though.

Also one other thing to consider (off the top of my head) is if the lists have dicts and the dicts have subdicts and those subdicts have different keys...

Reply:
> Also one other thing to consider (off the top of my head) is if the lists have dicts and the dicts have subdicts and those subdicts have different keys...

Yep, that won't be detected. It maybe possible to detect that but function would become extremely complex. It is already not easy to understand. Doing that would make it super complex but it is possible.

Reply:
Hm. Are you sure you don't want to go with the "simpler version" at this point? (I.e. the one you originally proposed of recursing fully and changing anything that looks like a float to a float?)

Reply:
No, this one's fine, too.